### PR TITLE
Replace Zuul with Spring Cloud Gateway for Reactive Microservices

### PR DESCRIPTION
**Pull request by mraible**
-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by PierreBesson** [comment]
@mraible An important consideration with Spring Cloud Gateway is that it only works on reactive applications. So we have to default to reactive for gateway type apps. We also have to drop support for Zuul entirely as it clearly not sustainable to support both technologies for the same functionality.

**Comment by mraible** [comment]
@PierreBesson I'm hoping to finish this integration in the next few weeks. My goal is to provide backward compatibility for JHipster 6.x users (by continuing to support Zuul) and to add support for reactive at the same time. We can remove the Zuul code as part of 7.x.

If 7.x was happening sooner, I'd remove Zuul. However, I'm guessing it's still a few months away.

**Comment by mraible** [comment]
I made some progress on the example today (see [this PR](https://github.com/mraible/jhipster-reactive-microservices-oauth2/pull/4) for more info). However, the access token doesn't seem to be getting forwarded to downstream services correctly.

**Comment by mraible** [comment]
I posted my issue as a question to Stack Overflow. https://stackoverflow.com/questions/60251863/spring-cloud-gateway-and-tokenrelay-filter

**Comment by mraible** [comment]
@jhipster/developers I have a [working prototype](https://github.com/mraible/jhipster-reactive-microservices-oauth2/pull/4) and I'm ready to finish this PR. However, I have a few questions before I do.

1. I removed `AuthorizedFeignClient`. It seems to be unused (by OAuth at least). Is this class necessary for another authentication implementation? There are some reactive Feign implementations available, but I'm not sure we should include it if it's not used.
2. I removed `GatewaySwaggerResourcesProvider`. It seems that Swagger is blocking #7608. If I leave it in the classpath, the app won't compile.
3. I removed `GatewayConfiguration`. This configures Swagger and access control. Access control seems to be handled by Spring Security now. I had to add `.pathMatchers("/services/**").authenticated()` for things to work.
4. I removed `GatewayResource`. Does this need to be restored with logic from Spring Cloud Gateway to return the list of routes? If so, do the routes need to come from the gateway or can they just be looked up from Eureka?

**Comment by mraible** [comment]
I fixed Swagger in https://github.com/mraible/jhipster-reactive-microservices-oauth2/pull/4/commits/42a5008664a9ebf8286f91eb53906195be98b4b7.

**Comment by mraible** [comment]
I posted [a question to Stack Overflow asking how to get route information](https://stackoverflow.com/questions/60269078/retrieving-routes-and-their-paths-from-spring-cloud-gateway-vs-zuul).

**Comment by mraible** [comment]
We should probably add a CI test for reactive microservices with OAuth and JWT (not implemented yet).

**Comment by mraible** [comment]
I generated a new project and `./mvnw verify` fails. Cause:

```
***************************
APPLICATION FAILED TO START
***************************

Description:

Parameter 0 of method gatewayMetricFilter in org.springframework.cloud.gateway.config.GatewayMetricsAutoConfiguration required a single bean, but 2 were found:
	- prometheusMeterRegistry: defined by method 'prometheusMeterRegistry' in class path resource [org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusMetricsExportAutoConfiguration.class]
	- consoleLoggingRegistry: defined by method 'consoleLoggingRegistry' in class path resource [io/github/jhipster/config/metric/JHipsterLoggingMetricsExportConfiguration.class]


Action:

Consider marking one of the beans as @Primary, updating the consumer to accept multiple beans, or using @Qualifier to identify the bean that should be consumed
```

**Comment by mraible** [comment]
Wahoo! All tests are passing. 🎉

The only issue I see now is the following when using a reactive monolith or gateway. I believe this is caused by the Swagger 3.0 upgrade.

```
2020-02-20 07:38:33.322 ERROR 57892 --- [  restartedMain] s.d.s.web.scanners.ApiDescriptionReader  : Skipping process path[/management/audits], method[getAll] as it has an error.

java.lang.NoSuchMethodError: 'com.google.common.collect.ImmutableSet springfox.documentation.spi.service.contexts.ParameterContext.getIgnorableParameterTypes()'
        at io.github.jhipster.config.apidoc.PageableParameterBuilderPlugin.createModelRefFactory(PageableParameterBuilderPlugin.java:211)
        at io.github.jhipster.config.apidoc.PageableParameterBuilderPlugin.createPageParameter(PageableParameterBuilderPlugin.java:154)
        at io.github.jhipster.config.apidoc.PageableParameterBuilderPlugin.apply(PageableParameterBuilderPlugin.java:107)
        at springfox.documentation.spring.web.plugins.DocumentationPluginsManager.operation(DocumentationPluginsManager.java:122)
        at springfox.documentation.spring.web.readers.operation.ApiOperationReader.read(ApiOperationReader.java:72)
        at springfox.documentation.spring.web.scanners.CachingOperationReader.lambda$new$0(CachingOperationReader.java:43)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at springfox.documentation.spring.web.scanners.CachingOperationReader.read(CachingOperationReader.java:48)
        at springfox.documentation.spring.web.scanners.ApiDescriptionReader.read(ApiDescriptionReader.java:72)
        at springfox.documentation.spring.web.scanners.ApiListingScanner.scan(ApiListingScanner.java:153)
        at springfox.documentation.spring.web.scanners.ApiDocumentationScanner.scan(ApiDocumentationScanner.java:69)
        at springfox.documentation.spring.web.plugins.AbstractDocumentationPluginsBootstrapper.scanDocumentation(AbstractDocumentationPluginsBootstrapper.java:97)
        at springfox.documentation.spring.web.plugins.AbstractDocumentationPluginsBootstrapper.bootstrapDocumentationPlugins(AbstractDocumentationPluginsBootstrapper.java:83)
        at springfox.documentation.spring.web.plugins.DocumentationPluginsBootstrapper.start(DocumentationPluginsBootstrapper.java:94)
        at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:182)
        at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:53)
        at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:360)
        at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:158)
        at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:122)
        at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:894)
        at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.finishRefresh(ReactiveWebServerApplicationContext.java:129)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:553)
        at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:66)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
        at io.github.jhipster.gateway.ReactiveGatewayApp.main(ReactiveGatewayApp.java:64)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49)

2020-02-20 07:38:33.346 ERROR 57892 --- [  restartedMain] s.d.s.web.scanners.ApiDescriptionReader  : Skipping process path[/management/audits], method[getByDates] as it has an error.

java.lang.NoSuchMethodError: 'com.google.common.collect.ImmutableSet springfox.documentation.spi.service.contexts.ParameterContext.getIgnorableParameterTypes()'
        at io.github.jhipster.config.apidoc.PageableParameterBuilderPlugin.createModelRefFactory(PageableParameterBuilderPlugin.java:211)
        at io.github.jhipster.config.apidoc.PageableParameterBuilderPlugin.createPageParameter(PageableParameterBuilderPlugin.java:154)
        at io.github.jhipster.config.apidoc.PageableParameterBuilderPlugin.apply(PageableParameterBuilderPlugin.java:107)
        at springfox.documentation.spring.web.plugins.DocumentationPluginsManager.operation(DocumentationPluginsManager.java:122)
        at springfox.documentation.spring.web.readers.operation.ApiOperationReader.read(ApiOperationReader.java:72)
        at springfox.documentation.spring.web.scanners.CachingOperationReader.lambda$new$0(CachingOperationReader.java:43)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at springfox.documentation.spring.web.scanners.CachingOperationReader.read(CachingOperationReader.java:48)
        at springfox.documentation.spring.web.scanners.ApiDescriptionReader.read(ApiDescriptionReader.java:72)
        at springfox.documentation.spring.web.scanners.ApiListingScanner.scan(ApiListingScanner.java:153)
        at springfox.documentation.spring.web.scanners.ApiDocumentationScanner.scan(ApiDocumentationScanner.java:69)
        at springfox.documentation.spring.web.plugins.AbstractDocumentationPluginsBootstrapper.scanDocumentation(AbstractDocumentationPluginsBootstrapper.java:97)
        at springfox.documentation.spring.web.plugins.AbstractDocumentationPluginsBootstrapper.bootstrapDocumentationPlugins(AbstractDocumentationPluginsBootstrapper.java:83)
        at springfox.documentation.spring.web.plugins.DocumentationPluginsBootstrapper.start(DocumentationPluginsBootstrapper.java:94)
        at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:182)
        at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:53)
        at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:360)
        at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:158)
        at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:122)
        at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:894)
        at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.finishRefresh(ReactiveWebServerApplicationContext.java:129)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:553)
        at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:66)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
        at io.github.jhipster.gateway.ReactiveGatewayApp.main(ReactiveGatewayApp.java:64)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49)

2020-02-20 07:38:33.687 ERROR 57892 --- [  restartedMain] s.d.s.web.scanners.ApiDescriptionReader  : Skipping process path[/api/users], method[getAllUsers] as it has an error.

java.lang.NoSuchMethodError: 'com.google.common.collect.ImmutableSet springfox.documentation.spi.service.contexts.ParameterContext.getIgnorableParameterTypes()'
        at io.github.jhipster.config.apidoc.PageableParameterBuilderPlugin.createModelRefFactory(PageableParameterBuilderPlugin.java:211)
        at io.github.jhipster.config.apidoc.PageableParameterBuilderPlugin.createPageParameter(PageableParameterBuilderPlugin.java:154)
        at io.github.jhipster.config.apidoc.PageableParameterBuilderPlugin.apply(PageableParameterBuilderPlugin.java:107)
        at springfox.documentation.spring.web.plugins.DocumentationPluginsManager.operation(DocumentationPluginsManager.java:122)
        at springfox.documentation.spring.web.readers.operation.ApiOperationReader.read(ApiOperationReader.java:72)
        at springfox.documentation.spring.web.scanners.CachingOperationReader.lambda$new$0(CachingOperationReader.java:43)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at springfox.documentation.spring.web.scanners.CachingOperationReader.read(CachingOperationReader.java:48)
        at springfox.documentation.spring.web.scanners.ApiDescriptionReader.read(ApiDescriptionReader.java:72)
        at springfox.documentation.spring.web.scanners.ApiListingScanner.scan(ApiListingScanner.java:153)
        at springfox.documentation.spring.web.scanners.ApiDocumentationScanner.scan(ApiDocumentationScanner.java:69)
        at springfox.documentation.spring.web.plugins.AbstractDocumentationPluginsBootstrapper.scanDocumentation(AbstractDocumentationPluginsBootstrapper.java:97)
        at springfox.documentation.spring.web.plugins.AbstractDocumentationPluginsBootstrapper.bootstrapDocumentationPlugins(AbstractDocumentationPluginsBootstrapper.java:83)
        at springfox.documentation.spring.web.plugins.DocumentationPluginsBootstrapper.start(DocumentationPluginsBootstrapper.java:94)
        at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:182)
        at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:53)
        at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:360)
        at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:158)
        at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:122)
        at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:894)
        at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.finishRefresh(ReactiveWebServerApplicationContext.java:129)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:553)
        at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:66)
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747)
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
        at io.github.jhipster.gateway.ReactiveGatewayApp.main(ReactiveGatewayApp.java:64)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49)
```

My `.yo-rc.json`:

```
{
  "generator-jhipster": {
    "applicationType": "gateway",
    "reactive": true,
    "baseName": "reactiveGateway",
    "packageName": "io.github.jhipster.gateway",
    "packageFolder": "io/github/jhipster/gateway",
    "authenticationType": "oauth2",
    "cacheProvider": "no",
    "enableHibernateCache": false,
    "websocket": false,
    "databaseType": "mongodb",
    "devDatabaseType": "mongodb",
    "prodDatabaseType": "mongodb",
    "searchEngine": false,
    "serviceDiscoveryType": "eureka",
    "buildTool": "gradle",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": ["en", "fr"],
    "testFrameworks": ["gatling", "protractor"],
    "serverPort": "8080",
    "jhiPrefix": "jhi",
    "clientPackageManager": "yarn",
    "clientFramework": "react",
    "jhipsterVersion": "6.7.1",
    "messageBroker": false,
    "embeddableLaunchScript": false,
    "useSass": true,
    "clientTheme": "none",
    "creationTimestamp": 1582178129646,
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "otherModules": [],
    "blueprints": []
  }
}
```

**Comment by mraible** [comment]
I started on the JWT implementation but ran into [issues](https://github.com/mraible/jhipster-reactive-microservices-jwt/pull/1).

**Comment by mraible** [comment]
The JWT implementation is working now! However, I did notice that I had to make the `secret.authentication.jwt.base64-secret` values match between the gateway and microservice. Is that expected?

File: generators/server/files.js [hunk 1]
```diff
@@ -543,7 +570,10 @@ const serverFiles = {
         },
         {
             condition: generator =>
-                generator.applicationType === 'gateway' && generator.authenticationType === 'jwt' && generator.serviceDiscoveryType,
+                !generator.reactive &&
+                generator.applicationType === 'gateway' &&
+                generator.authenticationType === 'jwt' &&
+                generator.serviceDiscoveryType,
```
Comment (mraible) [inline generators/server/files.js]: Why does Prettier not allow me to put everything on one line? It's annoying.

File: generators/server/files.js [hunk 2]
```diff
@@ -709,7 +739,10 @@ const serverFiles = {
         },
         {
             condition: generator =>
-                generator.authenticationType === 'oauth2' && generator.applicationType === 'gateway' && generator.serviceDiscoveryType,
+                !generator.reactive &&
+                generator.authenticationType === 'oauth2' &&
+                generator.applicationType === 'gateway' &&
+                generator.serviceDiscoveryType,
```
Comment (mraible) [inline generators/server/files.js]: Why does Prettier not allow me to put everything on one line? It's annoying.

**Comment by mraible** [comment]
I'm not sure why these tests are failing for JWT + Spring Cloud Gateway. They result in 500 errors instead of 200.

```
UserResourceIT > getAllAuthorities()
AuditResourceIT > getAllAudits()
```

[This commit](https://github.com/jhipster/generator-jhipster/pull/11223/commits/78e5f7f7e4c40535ca2f8184bfd5b4f37b749f94) disables them.

**Comment by Tcharl** [comment]
Hi @mraible ,
Sorry for being late on this, I just saw your comments.
`@AuthorizedFeignClient` is used by oauth configuration: it forwards the Authorization header when calling resource providers using Feign (backend to backend).

**Comment by mraible** [comment]
I thought it might. However, it doesn't seem to be used by default. Do you think it's OK to remove for reactive? 

If not, is there a reactive Feign client you recommend? The official one seems to be incubating.

https://github.com/OpenFeign/feign-reactive

This one might be better.

https://github.com/Playtika/feign-reactive

**Comment by cbornet** [comment]
I think for reactive we should support WebClient instead of Feign.

**Comment by mraible** [comment]
Should we pre-configure a WebClient or just leave it up to the user to do it? I'm hesitant to include it since it'll be unused by default. If we do want to include it, [Spring WebClient and OAuth2 Support](https://www.baeldung.com/spring-webclient-oauth2) seems to provide a good overview for OAuth.

**Comment by cbornet** [comment]
I'd say for now don't bother about WebClient. We can see later if there's demand for it.

**Comment by mraible** [comment]
@pascalgrimaud I re-enabled the failing tests in my last commit, but it didn't need to kick off a CI build. Can you please help?

**Comment by pascalgrimaud** [comment]
Ok, will have a look

**Comment by pascalgrimaud** [comment]
The last build didn't trigger yet, as you push to the organization instead your fork, maybe there are another builds in pending.

About the fix here https://github.com/jhipster/generator-jhipster/pull/11381/files, it's related to `testGetUnknownAccount`
Nothing related to your error:

```
[ERROR] Failures: 
[ERROR]   AuditResourceIT.getAllAudits:68 Status expected:<200 OK> but was:<500 INTERNAL_SERVER_ERROR>
```

```
[ERROR]   UserResourceIT.getAllAuthorities:462 Status expected:<200 OK> but was:<500 INTERNAL_SERVER_ERROR>
```

So for me, it's something different.

**Comment by mraible** [comment]
@pascalgrimaud I was able to fix the problem locally by switching from Couchbase to MongoDB. If [this commit](https://github.com/jhipster/generator-jhipster/pull/11223/commits/b2412959a3d85b8b941bde48ffef5d4772a7583a) fixes the issue, I'm inclined to ignore it and merge this PR. 

You said you have someone that's reviewing this PR. Is that still happening?

**Comment by pascalgrimaud** [comment]
yes, please @juliensadaoui you wanted to review it ?

**Comment by ruddell** [comment]
> The JWT implementation is working now! However, I did notice that I had to make the `secret.authentication.jwt.base64-secret` values match between the gateway and microservice. Is that expected?

When using the JHipster Registry, the JWT secret is fetched from the Cloud Config, so all apps end up using the same value.  Apps use the [`secret.authentication.jwt.secret`](https://github.com/jhipster/generator-jhipster/blob/9e338dc9b9e496eb4cbadc8b0c0735ff0c975660/generators/server/templates/src/main/docker/config/localhost-config/application.yml.ejs#L10) provided by the registry based on [this code](https://github.com/jhipster/generator-jhipster/blob/9e338dc9b9e496eb4cbadc8b0c0735ff0c975660/generators/server/templates/src/main/java/package/security/jwt/TokenProvider.java.ejs#L65-L72).  Does Reactive + Spring Cloud Gateway have a centralized config source?

**Comment by mraible** [comment]
@pascalgrimaud Can I assign a bounty to a code review? 😆

**Comment by mraible** [comment]
> Does Reactive + Spring Cloud Gateway have a centralized config source?

@ruddell I was using the same Eureka + Spring Cloud Config option that our Zuul gateway uses. I also tried Consul and confirmed it works.

**Comment by pascalgrimaud** [comment]
@mraible : of course ! you're our main sponsor with Okta, you can !!!

**Comment by mraible** [comment]
@pascalgrimaud OK, I assigned a bug bounty to this PR. Whoever does the code review gets it!

**Comment by pascalgrimaud** [comment]
For the one who will review, it's important to check our default Microservices options to be sure there is no regression.
Then, of course, test the reactive part

**Comment by mraible** [comment]
The entity generator isn't enabled for reactive so it's difficult to test communication between reactive microservices. That's why I created these samples that communicate with a downstream microservice.

* https://github.com/mraible/jhipster-reactive-microservices-jwt
* https://github.com/mraible/jhipster-reactive-microservices-oauth2

Each has a single PR in a `reactive` branch so you can see what's changed.

I just re-generated both gateways in them today.

**Comment by mraible** [comment]
@pascalgrimaud I'm tempted to merge this since no reviewers seem to be coming forward.

**Comment by pascalgrimaud** [comment]
@mraible : @juliensadaoui said to me he already started to review and test it, but didn't finish

**Comment by pascalgrimaud** [comment]
Just merged this https://github.com/jhipster/generator-jhipster/pull/11417
Can you resolve the conflict plz @mraible ?

**Comment by mraible** [comment]
@pascalgrimaud Conflicts have been resolved.

**Comment by mraible** [comment]
If we do decide to pre-configure a `WebClient` for OAuth 2.0 in the future, [there's a good how-to on Stack Overflow](https://stackoverflow.com/questions/58982286/spring-security-5-replacement-for-oauth2resttemplate/58999002#58999002).

**Comment by mraible** [comment]
This release of Spring Cloud should allow us to get rid of the Spring snapshot repos. https://spring.io/blog/2020/03/05/spring-cloud-hoxton-service-release-3-sr3-is-available

**Comment by DanielFran** [comment]
@mraible Will update bom tonight

**Comment by mraible** [comment]
@DanielFran I updated to Hoxton SR3 in https://github.com/jhipster/jhipster/pull/611.

**Comment by mraible** [comment]
Here are the steps I'll be using before merging in the near future.

. Use my blog + store microservices example to generate a Spring MVC example, verify `yarn e2e` passes after waiting 10+ minutes for Docker to start everything
. Use my blog + store microservices example to generate a WebFlux example, verify `yarn e2e` passes

If both work, I'm merging. Speak now or create a new PR to fix. :)

**Comment by pascalgrimaud** [comment]
Ok, I try to find time to do the same. Can you just post your JDL about Reactive ?

**Comment by mraible** [comment]
```
application {
  config {
    baseName gateway,
    packageName com.okta.developer.gateway,
    applicationType gateway,
    authenticationType oauth2,
    databaseType couchbase,
    devDatabaseType couchbase,
    prodDatabaseType couchbase,
    serviceDiscoveryType eureka,
    testFrameworks [protractor],
  }
  entities Blog, Post, Tag, Product
}

application {
  config {
    baseName blog,
    packageName com.okta.developer.blog,
    applicationType microservice,
    authenticationType oauth2,
    databaseType mongodb,
    devDatabaseType mongodb,
    prodDatabaseType mongodb,
    serverPort 8081,
    serviceDiscoveryType eureka
  }
  entities Blog, Post, Tag
}

application {
  config {
    baseName store,
    packageName com.okta.developer.store,
    applicationType microservice,
    authenticationType oauth2,
    databaseType mongodb,
    devDatabaseType mongodb,
    prodDatabaseType mongodb,
    enableHibernateCache false,
    searchEngine elasticsearch,
    serverPort 8082,
    serviceDiscoveryType eureka,
  }
  entities Product
}

entity Blog {
  name String required minlength(3),
  handle String required minlength(2)
}

entity Post {
  title String required,
  content TextBlob required,
  date Instant required
}

entity Tag {
  name String required minlength(2)
}

entity Product {
  title String required,
  price BigDecimal required min(0),
  image ImageBlob
}

relationship ManyToOne {
  Blog{user(login)} to User,
  Post{blog(name)} to Blog
}

relationship ManyToMany {
  Post{tag(name)} to Tag{post}
}

paginate Post, Tag with infinite-scroll
paginate Product with pagination

microservice Product with store
microservice Blog, Post, Tag with blog

// will be created under 'docker-compose' folder
deployment {
  deploymentType docker-compose
  appsFolders [gateway, blog, store]
  dockerRepositoryName "mraible"
  monitoring elk
  consoleOptions [zipkin]
}

// will be created under 'kubernetes' folder
deployment {
  deploymentType kubernetes
  appsFolders [gateway, blog, store]
  dockerRepositoryName "mraible"
  monitoring elk
  consoleOptions [zipkin]
}
```

**Comment by mraible** [comment]
I realized this test is not fair. Since this is a new feature, I'll be testing the `master` branch's output from the JDL above versus the JDL from the `reactive-gateway` branch. If they match, I'm merging! 👯‍♂

**Comment by pascalgrimaud** [comment]
yes go ahead and merge if you feel confident
I can open issue later if I find something

My goal is to do a release this monday. Is it ok for your next talks ?

**Comment by mraible** [comment]
@pascalgrimaud Thank you! I feel confident but there's a lot of moving parts too. My talks are in flux, but I can always do live streams on YouTube. Thanks for all your help. You're awesome!! 😃🤘

**Comment by pascalgrimaud** [comment]
Good, thanks.
So I keep my target on Monday, for the release !

**Pull request merged**